### ISSUE TITLE
custom optimizer in pure r to pass into keras

### ISSUE THREAD

**Issue by ds5j** [rstudio/keras#513]
Hi,
   Is it possible to create your own custom optimizer algorithm in pure r and pass it into the optimizer function?

Does the keras API expect certain outputs from optimizer class?  Since it is calling self is it just calling a python optimizer?

thanks

**Comment by skeydan** [comment]
I guess this won't be easy if you're using the usual Keras sequential

```define model -> compile -> fit```

workflow. 

See 

https://github.com/jjallaire/deep-learning-with-r-notebooks/blob/master/notebooks/8.3-neural-style-transfer.Rmd 

(from @jjallaire Deep Learning with R) 

though for a case where you're obtaining the losses and gradients yourself, and then, use R ```optim``` for optimization.

If you use the eager execution workflow detailed in 

https://blogs.rstudio.com/tensorflow/posts/2018-07-30-attention-layer/

you're already doing all this (get loss, get gradients, apply gradients) step by step, so with that workflow it should be less complicated to add your own optimization algorithm.
